### PR TITLE
Simple path tab-completion.

### PR DESCRIPTION
**Pull request by kubaroth**
Simple path tab-completion, matching first selection. Appending trailing / to minimize keystrokes with subsequent completions.

### CODE REVIEW

**Comment by kubaroth** [comment]
oops, will fix that today.

**Comment by kubaroth** [comment]
hmmm, 406.1 is still failing about some missing methods. I'm having a hard time to find any documentation older then qt5.5 at the moment to check when those features were introduced. Perhaps I should change the support from 5.5 up? What do you think?

File: fm/src/mainwindow.cpp [hunk 1]
```diff
@@ -152,6 +152,9 @@ MainWindow::MainWindow()
     dockBookmarks->setSizePolicy(QSizePolicy::Fixed,QSizePolicy::Fixed);
     bookmarksList = new QListView(dockBookmarks);
     bookmarksList->setMinimumHeight(24); // Docks get the minimum size from their content widget
+#if QT_VERSION >= 0x050500
```
Comment (rodlie) [inline fm/src/mainwindow.cpp]:
```suggestion

```

File: fm/src/mainwindow.cpp [hunk 2]
```diff
@@ -152,6 +152,9 @@ MainWindow::MainWindow()
     dockBookmarks->setSizePolicy(QSizePolicy::Fixed,QSizePolicy::Fixed);
     bookmarksList = new QListView(dockBookmarks);
     bookmarksList->setMinimumHeight(24); // Docks get the minimum size from their content widget
+#if QT_VERSION >= 0x050500
+     bookmarksList->setFocusPolicy(Qt::FocusPolicy::NoFocus); // Avoid hijacking focus when Tab on Edit Path
```
Comment (rodlie) [inline fm/src/mainwindow.cpp]:
```suggestion
     bookmarksList->setFocusPolicy(Qt::NoFocus); // Avoid hijacking focus when Tab on Edit Path
```

File: fm/src/mainwindow.cpp [hunk 3]
```diff
@@ -152,6 +152,9 @@ MainWindow::MainWindow()
     dockBookmarks->setSizePolicy(QSizePolicy::Fixed,QSizePolicy::Fixed);
     bookmarksList = new QListView(dockBookmarks);
     bookmarksList->setMinimumHeight(24); // Docks get the minimum size from their content widget
+#if QT_VERSION >= 0x050500
+     bookmarksList->setFocusPolicy(Qt::FocusPolicy::NoFocus); // Avoid hijacking focus when Tab on Edit Path
+#endif
```
Comment (rodlie) [inline fm/src/mainwindow.cpp]:
```suggestion

```

File: fm/src/mainwindow.cpp [hunk 4]
```diff
@@ -1621,6 +1624,32 @@ bool MainWindow::eventFilter(QObject *o, QEvent *e)
         default:;
         }
     }
+#if QT_VERSION >= 0x050500
```
Comment (rodlie) [inline fm/src/mainwindow.cpp]:
```suggestion
```

File: fm/src/mainwindow.cpp [hunk 5]
```diff
@@ -1621,6 +1624,32 @@ bool MainWindow::eventFilter(QObject *o, QEvent *e)
...
+        if (e->type()==QEvent::KeyPress) {
+            QKeyEvent* key = static_cast<QKeyEvent*>(e);
+            if ( (key->key()==Qt::Key_Tab) ) {
+                qDebug()<< "Tab pressed: path completion "<< o ;
+                QListView *completionList = dynamic_cast<QListView*>(o);
+                // Remove incomplete phrase and replace it with a current index
+                QModelIndex index = completionList->currentIndex();
+                QString itemText = index.data(Qt::DisplayRole).toString();
+                QString currentPath = pathEdit->lineEdit()->text();
+                QList<QString> tempList = currentPath.split('/');
```
Comment (rodlie) [inline fm/src/mainwindow.cpp]:
```suggestion
                QStringList tempList = currentPath.split("/");
```

File: fm/src/mainwindow.cpp [hunk 6]
```diff
@@ -1621,6 +1624,32 @@ bool MainWindow::eventFilter(QObject *o, QEvent *e)
...
+                tempList.takeLast();
+                tempList.push_back(itemText);
```
Comment (rodlie) [inline fm/src/mainwindow.cpp]:
```suggestion
                tempList << itemText;
```

File: fm/src/mainwindow.cpp [hunk 7]
```diff
@@ -1621,6 +1624,32 @@ bool MainWindow::eventFilter(QObject *o, QEvent *e)
...
+                QString newPath = tempList.join('/');
```
Comment (rodlie) [inline fm/src/mainwindow.cpp]:
```suggestion
                QString newPath = tempList.join("/");
```

File: fm/src/mainwindow.cpp [hunk 8]
```diff
@@ -1621,6 +1624,32 @@ bool MainWindow::eventFilter(QObject *o, QEvent *e)
...
+                pathEdit->lineEdit()->setText(newPath);
+                // Force update the Main View
+                emit pathEdit->activated(newPath);
```
Comment (rodlie) [inline fm/src/mainwindow.cpp]:
```suggestion
                pathEditChanged(newPath);
```

File: fm/src/mainwindow.cpp [hunk 9]
```diff
@@ -1621,6 +1624,32 @@ bool MainWindow::eventFilter(QObject *o, QEvent *e)
...
+                // Enter Edit Mode right after the event system is ready
+                QTimer::singleShot(0, pathEdit->lineEdit(), SLOT(setFocus()));
+                // Add the trailing / for subsequent completions
+                pathEdit->lineEdit()->setText(newPath + QString("/"));
+            }
+        }
+    }
+#endif
```
Comment (rodlie) [inline fm/src/mainwindow.cpp]:
```suggestion

```

**Comment by kubaroth** [comment]
sweet, thanks for taking a look.

**Pull request merged**